### PR TITLE
style: lint and styling overhaul

### PR DESCRIPTION
**Pull request by Benjozork**
## Summary of Changes

* Bump eslint-related package versions
* Consolidate eslint configs for modern TS code under one config
* Introduce Prettier as main formatting style, integrated with ESLint
* Fix linting of package base JS
* Fix lint issues now coming up after cleanup of configs/upgrade of packages

## Testing instructions

- Should not affect code, but make sure nothing is obviously broken
- Focus on that the guidance (LNAV/VNAV) does not have any new bugs 
  - Routes (or similar):
    - EDDM-LOWI LOC26R (circle to land) 
    - KCRQ24 PADRZ2 CHKNN AVE ROBIE5 BORED KSJC30L (ILS)
    - EDDF/25L F250 SULUS1F SULUS Z12 LASGA T202 ATGUP ATG25L EDDB/25L
    EBBR-EGKK26L 
- EFB does not have incorrect/broken formatting

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by tracernz — APPROVED** [review]

**Review by tracernz — APPROVED** [review]
Still a couple of lint errors left to clean up.

**Comment by tracernz** [comment]
Did a full flight (NZTG -> NZWN). No issues to report.

Probably should get a smoke test after the conflicts are solved looking at the size of them.

File: fbw-a32nx/src/systems/instruments/buildSrc/aceBuild.mjs
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/buildSrc/aceBuild.mjs:1]: Is this intended?

File: fbw-a32nx/src/systems/instruments/buildSrc/plugins.mjs
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/buildSrc/plugins.mjs:1]: ^

File: fbw-a32nx/src/systems/tcas/src/components/TcasComputer.ts [hunk 1]
```diff
@@ -7,1189 +7,1304 @@
...
+              traffic.secondsSinceLastTa = 0;
+              break;
+            default:
+              if (traffic.secondsSinceLastTa >= TCAS.TA_EXPIRATION_DELAY) {
+                traffic.secondsSinceLastTa = 0;
+                traffic.intrusionLevel = desiredIntrusionLevel;
+                if (this.debug)
+                  console.log(
+                    `${traffic.ID} resetting seconds since last TA to ${traffic.secondsSinceLastTa} - new intrusion level is ${desiredIntrusionLevel}`,
+                  );
```
Comment (2hwk) [inline fbw-a32nx/src/systems/tcas/src/components/TcasComputer.ts:684]:
```suggestion
                if (this.debug) {
                  console.log(
                    `${traffic.ID} resetting seconds since last TA to ${traffic.secondsSinceLastTa} - new intrusion level is ${desiredIntrusionLevel}`,
                  );
                }
```

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/FmsDataPublisher.ts [hunk 2]
```diff
@@ -1,50 +1,66 @@
...
+  constructor(bus: EventBus, stateSubject: Subject<'L' | 'R'>) {
+    super(
+      new Map([
+        [
+          'ndMessageFlags',
+          { name: (side) => `L:A32NX_EFIS_${side}_ND_FM_MESSAGE_FLAGS`, type: SimVarValueType.Number },
+        ],
+        ['crossTrackError', { name: (_side) => 'L:A32NX_FG_CROSS_TRACK_ERROR', type: SimVarValueType.NM }],
+        [
+          'linearDeviationActive',
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/FmsDataPublisher.ts:38]: Personally, I dislike how the object new lining is handled here, why not be consistent through the entire map and split all the objects? There isn't any major character saving either, just makes working with the map slightly more annoying.

File: fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/FuelSystemPublisher.ts [hunk 3]
```diff
@@ -31,23 +49,40 @@ type FuelSystemIndexedEvents = {
...
+      ],
+      [
+        'fuel_engine_pressure',
+        { name: 'FUELSYSTEM ENGINE PRESSURE:#index#', type: SimVarValueType.PSI, indexed: true },
+      ],
+      [
+        'fuel_line_pressure',
+        { name: 'FUELSYSTEM LINE FUEL PRESSURE:#index#', type: SimVarValueType.PSI, indexed: true },
+      ],
+      ['fuel_line_flow', { name: 'FUELSYSTEM LINE FUEL FLOW:#index#', type: SimVarValueType.GPH, indexed: true }],
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/FuelSystemPublisher.ts:82]: Same as above

File: fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/SwitchableProvider.ts [hunk 4]
```diff
@@ -1,31 +1,43 @@
...
+          k,
+          {
             name: v.name(stateSubject.get()),
             type: v.type,
-        }])), bus);
+          },
+        ]),
+      ),
+      bus,
+    );
```
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/SwitchableProvider.ts:31]: This is a lot less readable after the linting change

File: fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/TcasBusPublisher.ts [hunk 5]
```diff
@@ -1,17 +1,24 @@
...
-    }
+  constructor(bus: EventBus) {
+    super(
+      new Map([
+        ['tcasTaOnly', { name: 'L:A32NX_TCAS_TA_ONLY', type: SimVarValueType.Bool }],
+        ['tcasFault', { name: 'L:A32NX_TCAS_FAULT', type: SimVarValueType.Bool }],
+        ['tcasMode', { name: 'L:A32NX_TCAS_MODE', type: SimVarValueType.Number }],
+      ]),
+      bus,
+    );
```
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/TcasBusPublisher.ts:22]: Ditto

File: fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/VorBusPublisher.ts [hunk 6]
```diff
@@ -1,134 +1,141 @@
...
-            ['nav1RelativeBearing', { name: 'NAV RELATIVE BEARING TO STATION:1', type: SimVarValueType.Degree }],
-            ['nav1Obs', { name: 'NAV OBS:1', type: SimVarValueType.Degree }],
-            ['nav1Localizer', { name: 'NAV LOCALIZER:1', type: SimVarValueType.Degree }],
-            ['nav1RadialError', { name: 'NAV RADIAL ERROR:1', type: SimVarValueType.Degree }],
-            // This is fine at the moment. The database does not have the necessary information to implement proper logic.
-            ['nav1Available', { name: 'NAV HAS NAV:1', type: SimVarValueType.Bool }],
-            ['nav1StationDeclination', { name: 'NAV MAGVAR:1', type: SimVarValueType.Degree }],
-            ['nav1Location', { name: 'NAV VOR LATLONALT:1', type: SimVarValueType.LLA }],
+  constructor(bus: EventBus) {
+    super(
```
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/MsfsAvionicsCommon/providers/VorBusPublisher.ts:73]: Ditto

File: fbw-a32nx/src/systems/instruments/src/PFD/AttitudeIndicatorFixed.tsx [hunk 7]
```diff
@@ -12,515 +12,602 @@ import { Arinc429Values } from './shared/ArincValueProvider';
...
+        -16,
+      ) * -1.43875;
+    const totalRollCommand =
+      Math.max(
+        Math.min(
+          (foStickDisabled ? 0 : this.foRollCommand.value) + (captStickDisabled ? 0 : this.captRollCommand.value),
+          20,
+        ),
+        -20,
+      ) * 1.478;
```
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/PFD/AttitudeIndicatorFixed.tsx:510]: Expanded like this it is a bit hard to follow. Maybe more because of the magic numbers.  Potentially add a TODO here.

File: fbw-a32nx/src/systems/instruments/src/PFD/AttitudeIndicatorHorizon.tsx [hunk 8]
```diff
@@ -21,581 +37,732 @@ const DisplayRange = 35;
...
+            </text>
+            <text x="96.090324" y="140.69786">
+              50
+            </text>
+            <text x="43.125065" y="170.80962">
+              80
+            </text>
+            <text x="101.38947" y="170.80959">
+              80
+            </text>
```
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/PFD/AttitudeIndicatorHorizon.tsx:304]: The whitespace addition here is a bit much...

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/WeatherWidget.tsx [hunk 9]
```diff
@@ -4,316 +4,307 @@
...
-                {metar.barometer.mb.toFixed(0)}
-                {' '}
-                mb
-            </>
-        );
-    };
+  const BaroValue = () => {
+    const displayedBaroType = baroType === 'AUTO' ? getBaroTypeForAirport(metar.icao) : baroType;
+    if (displayedBaroType === 'IN HG') {
+      return <>{metar.barometer.hg.toFixed(2)} inHg</>;
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/WeatherWidget.tsx:103]: While this is more concise it is also easier to exclude the space in future changes, just note this.

File: fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/WeatherWidget.tsx [hunk 10]
```diff
@@ -4,316 +4,307 @@
...
+  };
 
-    async function getMetar(icao: string, source: string): Promise<void> {
-        if (icao.length !== 4 || !(/^[a-z]{4}$/i.test(icao))) {
-            setErrorMetar(t('Dashboard.ImportantInformation.Weather.NoIcaoProvided'));
-            dispatch(setMetar(MetarParserTypeProp));
-            return Promise.resolve();
-        }
+  async function getMetar(icao: string, source: string): Promise<void> {
+    if (icao.length !== 4 || !/^[a-z]{4}$/i.test(icao)) {
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/WeatherWidget.tsx:123]: I prefer wrapping regex in () personally

File: fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/FlightWidget.tsx [hunk 11]
```diff
@@ -9,233 +9,254 @@ import { CloudArrowDown } from 'react-bootstrap-icons';
...
+                  <div className="bg-theme-accent mx-4 my-auto h-8 w-1" />
+                  <InformationEntry title={t('Dashboard.YourFlight.CruiseAlt')} info={crzAlt} />
+                </div>
+              </div>
+              <div>
+                <h5 className="mb-2 text-2xl font-bold">{t('Dashboard.YourFlight.Route')}</h5>
+                <ScrollableContainer height={15}>
+                  <p className="font-mono text-2xl">
+                    <span className="text-theme-highlight text-2xl">
+                      {departingAirport}/{departingRunway}
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/FlightWidget.tsx:220]: Really lost at trying to decipher what the linter thinks its doing here... 

Substantially less readable imo.

File: fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/ColorMetar.tsx [hunk 12]
```diff
@@ -9,37 +9,34 @@ import { ColorCode, MetarParserType } from '@flybywiresim/fbw-sdk';
...
+      case ColorCode.TrendMarker:
+        style = 'font-bold text-gray-500';
+        break;
+      default:
+    }
 
     return (
-        <span>{ partsList }</span>
+      <span key={metarPart} className={`text-2xl ${style}`}>
+        {metarPart}{' '}
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Dashboard/Widgets/ColorMetar.tsx:36]: I don't understand either the before or after, equally bad. Manually clean this up at some point...

File: fbw-common/src/typings/asobo-vcockpits-instruments-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.d.ts [hunk 13]
```diff
@@ -2,37 +2,59 @@
 
 declare global {
```
Comment (2hwk) [inline fbw-common/src/typings/asobo-vcockpits-instruments-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/EICAS/ECAM/A320_Neo_ECAMGauge.d.ts:3]: Not sure if typing/*.d.ts should be linted

File: fbw-common/src/systems/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeedAuto/GroundSpeedAuto.tsx [hunk 14]
```diff
@@ -9,53 +9,53 @@ import { removeTodGroundSpeed, setTodGroundSpeed, setTodGroundSpeedMode } from '
...
+
+  return (
+    <div className="flex flex-col items-center justify-center">
+      <div className="bg-theme-accent absolute top-0 flex w-full flex-row items-center space-x-4 p-2">
+        <div className="bg-theme-highlight h-6 w-6 shrink-0 rounded-full" />
+        <p>{t('Performance.TopOfDescent.GroundSpeed.FetchingSpeedFromSimulator')}</p>
+      </div>
+
+      <span className="mb-4 text-5xl">
+        {groundSpeed || 0} {t('Performance.TopOfDescent.GroundSpeed.UnitKt')}
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/TODCalculator/GroundSpeed/GroundSpeedAuto/GroundSpeedAuto.tsx:49]: same issue with non-breaking space as before for readability

File: fbw-common/src/systems/instruments/src/EFB/Store/features/navigationPage.ts [hunk 15]
```diff
@@ -8,205 +8,209 @@ import { store, RootState } from '../store';
...
+      state.pinnedCharts[editIndex] = { ...state.pinnedCharts[editIndex], ...editedProperties };
     },
+  },
 });
 
 /**
  * @returns Whether or not associated chart with the passed chartId is pinned
  */
-export const isChartPinned = (chartId: string): boolean => (store.getState() as RootState).navigationTab.pinnedCharts.some((pinnedChart) => pinnedChart.chartId === chartId);
+export const isChartPinned = (chartId: string): boolean =>
```
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Store/features/navigationPage.ts:203]: enclose with {}

File: fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 16]
```diff
@@ -12,495 +12,602 @@ import { Spoilers } from '../../Common/Spoilers';
...
-        doorNotLockedUp = lgciu1DiscreteWord1.getBitValue(17) || lgciu2DiscreteWord1.getBitValue(17);
+const GearDoor = ({
+  x,
+  y,
+  location,
+  lgciu1DiscreteWord1,
+  lgciu2DiscreteWord1,
+  lgciu1DiscreteWord3,
+  lgciu2DiscreteWord3,
+}: GearDoorProps) => {
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx:284]: 🧐

File: fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 17]
```diff
@@ -12,495 +12,602 @@ import { Spoilers } from '../../Common/Spoilers';
...
-    );
+const LandingGearPositionIndicators = ({
+  x,
+  y,
+  location,
+  lgciu1DiscreteWord1,
+  lgciu2DiscreteWord1,
+  lgciu1DiscreteWord3,
+  lgciu2DiscreteWord3,
+}: LandingGearPositionIndicatorsProps) => {
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx:410]: ^

File: fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx [hunk 18]
```diff
@@ -12,495 +12,602 @@ import { Spoilers } from '../../Common/Spoilers';
...
-    </SvgGroup>
+const Gear = ({
+  x,
+  y,
+  location,
+  lgciu1DiscreteWord1,
+  lgciu2DiscreteWord1,
+  lgciu1DiscreteWord3,
+  lgciu2DiscreteWord3,
+}: GearProps) => (
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/SD/Pages/Wheel/Wheel.tsx:517]: ^

File: fbw-a32nx/src/systems/instruments/src/SD/PagesContainer.tsx [hunk 19]
```diff
@@ -42,219 +42,237 @@ const FCTL_CONDITION_TIMER_DURATION = 20;
...
-                checkEnginePage(deltaTime);
-                break;
-            case 2:
-                const sidestickPosX = SimVar.GetSimVarValue('L:A32NX_SIDESTICK_POSITION_X', 'Number');
-                const sidestickPosY = SimVar.GetSimVarValue('L:A32NX_SIDESTICK_POSITION_Y', 'Number');
-                const rudderPos = SimVar.GetSimVarValue('RUDDER PEDAL POSITION', 'Position');
-                const controlsMoved = Math.abs(sidestickPosX) > 0.05 || Math.abs(sidestickPosY) > 0.05 || Math.abs(rudderPos) > 0.2;
+      switch (fwcFlightPhase) {
+        case 10:
+        case 1:
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/SD/PagesContainer.tsx:149]: What are the rules for adding brackets {} on switch statements? Seems a bit inconsistent to only wrap a case in them if you have a variable involved.

File: fbw-a32nx/src/systems/instruments/src/SD/Text/Text.tsx [hunk 20]
```diff
@@ -1,41 +1,44 @@
...
-
-    if (title) {
-        return produceTextWithClass('title');
-    }
-
-    if (warning) {
-        return produceTextWithClass('warning');
-    }
-
+  const { x, y, unit, bigValue, value, title, warning, alignEnd, alignStart, children } = props;
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/SD/Text/Text.tsx:9]: Similar size object destructures are getting split, but not this one

File: fbw-a32nx/src/systems/instruments/src/util.js [hunk 21]
```diff
@@ -1,107 +1,111 @@
...
+  });
 }
 
 const SIMVAR_TYPES = {
-    '__proto__': null,
-    'GPS POSITION LAT': 'degrees latitude',
-    'L:A32NX_COLD_AND_DARK_SPAWN': 'Bool',
-    'TOTAL AIR TEMPERATURE': 'Celsius',
-    'AMBIENT TEMPERATURE': 'Celsius',
+  __proto__: null,
```
Comment (alepouna) [inline fbw-a32nx/src/systems/instruments/src/util.js:53]: Is this intended to be unquoted? Concern regarding how this is handled, especially if the intention is to not set the object prototype

**Comment by marvin-wtt** [comment]
I think most comments here regarding the code style are not about eslint but more about prettier. Maybe add it to the PR workflows to have a common style?

File: fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/EWD/PseudoFWC.ts:1]: Good candidate to exclude from this styling change entirely. Will be replaced by mostly by the new FWC.

File: fbw-a32nx/src/systems/tcas/src/components/TcasComputer.ts [hunk 22]
```diff
@@ -7,1189 +7,1304 @@
...
+      this.radioAlt.isFailureWarning() ||
+      (!this.radioAlt.isNoComputedData() && this.radioAlt.value < 500) ||
+      this.gpwsWarning ||
+      this.tcasMode.getVar() === TcasMode.STBY
+    ) {
+      this.inhibitions = Inhibit.ALL_RA_AURAL_TA;
+    } else if (
+      (!this.radioAlt.isNoComputedData() && this.radioAlt.value < 1000) ||
+      this.tcasMode.getVar() === TcasMode.TA
+    ) {
```
Comment (2hwk) [inline fbw-a32nx/src/systems/tcas/src/components/TcasComputer.ts:338]: ehhh

File: fbw-common/src/systems/datalink/common/src/com/FutureAirNavigationSystem.ts
Comment (2hwk) [inline fbw-common/src/systems/datalink/common/src/com/FutureAirNavigationSystem.ts:1]: Another one that should be excluded

File: fbw-common/src/systems/datalink/common/src/components/UplinkMonitor.ts [hunk 23]
```diff
@@ -5,167 +5,184 @@ import { CpdlcMessage } from '../messages/CpdlcMessage';
...
+    'UM25',
+    'UM65',
+    'UM77',
+    'UM83',
+    'UM84',
+    'UM97',
+    'UM118',
+    'UM121',
+    'UM130',
+  ];
```
Comment (2hwk) [inline fbw-common/src/systems/datalink/common/src/components/UplinkMonitor.ts:26]: Expanded list

**Review by 2hwk — CHANGES_REQUESTED** [review]

**Comment by 2hwk** [comment]
> I think most comments here regarding the code style are not about eslint but more about prettier. Maybe add it to the PR workflows to have a common style?

Good point, I'll modify a few of the comments, "linter" isn't very correct

File: fbw-a32nx/src/systems/instruments/src/SD/PagesContainer.tsx (hunk 19 above)
Comment (2hwk) [reply fbw-a32nx/src/systems/instruments/src/SD/PagesContainer.tsx:149]: I think it's to do with another linting rule about not declaring variables in switch case

**Comment by Benjozork** [comment]
After analysis the `__proto__` change is only used in dead code, which we will remove in another PR

**Review by 2hwk — APPROVED** [review]

**Comment by 2hwk** [comment]
This will enforce a merge freeze on ts/js/tsx when we are prepared on our side. At this moment, the status of this is pending.

**Comment by 2hwk** [comment]
Quality Assurance Report

Tier of Testing  : 2
Date  : 27/04/2024

Testing Process: 
- [x] Route YPPH/24 N0429F370 AVNEX3 AVNEX Y15 ESDEG/N0433F390 Y15 VETEL Z55 KA DCT YPKA/08 
- [x] Checked for bugs with LNAV/VNAV - Found but consistent with master
- [ ] EFB check not passed

Negatives:
- [ ] EFB regression - Pushback page indicator no longer works reliably.

Testing Results: 
Failed

Conclusions: 
Not Passed.

**Comment by 2hwk** [comment]
Quality Assurance Report

Tier of Testing  : 2
Date  : 27/04/2024

Testing Process: 
- [x] Route KCRQ24 PADRZ2 CHKNN AVE ROBIE5 BORED KSJC30L (ILS)
- [x] LNAV/VNAV - Bug found with SID entry removing runway but consistent with master

Negatives:
- N/A

Testing Results: 
Passed

Conclusions: 
Passed
